### ISSUE TITLE
Add support for handling returned (undeliverable) messages in Sender

### ISSUE THREAD

**Issue by a701440** [reactor/reactor-rabbitmq#95]
Looks like functionality of installing a Channel.addReturnListener is missing.
Sender.sendWithPublishConfirms does not handle undeliverable messages.

**Comment by acogoluegnes** [comment]
This should be doable by adding an `trackReturtned` option to `SendOptions` and a `returned` boolean property to `OutboundMessageResult` (naming is not definitive).

**Comment by vanseverk** [comment]
@acogoluegnes can I take this issue onto me?

**Comment by acogoluegnes** [comment]
@vanseverk Sure, thanks!

**Comment by vanseverk** [comment]
@acogoluegnes I seem to be running into an issue using the `returnListener` in the `Sender` class.

It has a method with the signature `handleReturn(int replyCode, String replyText, String exchange, String routingKey, AMQP.BasicProperties properties, byte[] body)` but does not seem to return any deliveryTag. 

This means the  unconfirmed message can not be retrieved, won't be passed along the `OutboundMessageResult` and won't be deleted from the `ConcurrentNavigableMap<Long, OutboundMessage> unconfirmed`.

**Comment by acogoluegnes** [comment]
It is possible to create a `OutboundMessage` from the values in `handleReturn` and wrap it in a `OutboundMessageResult` to emit it downstream.

Unfortunately, this would not help in cleaning the `unconfirmed` map with messages that could not be routed. I'd suggest adding the delivery tag into a custom `reactor_rabbitmq_delivery_tag` header to be able to get it back in the return listener. @vanseverk WDYT?

**Comment by vanseverk** [comment]
@acogoluegnes well, adding the delivery tag in a custom header might be a solution to the problem. Would this not mean that we're sending a duplicate one, however? Assuming the current delivery tag is already in a header. 

In case we can add and succesfully retrieve (or signal an error in case it's missing) the header, I believe this would indeed be the solution I was considering myself as well.

**Comment by acogoluegnes** [comment]
The delivery tag does not travel on the wire, it is just a sequence the client and the server increment independently, so there is no duplication here.

**Comment by acogoluegnes** [comment]
@vanseverk Maybe you'll want to rebase your work on master because of #99, which provided a fix to remove the confirm listener at the end of the flow. This is useful when channel pooling is in use. Note the return listener should be removed as well.

**Comment by vanseverk** [comment]
@acogoluegnes thanks for the heads-up!

**Comment by vanseverk** [comment]
@acogoluegnes it seems by default the BasicProperties object in OutboundMessage is "null" by default, and the headers inside of it are empty as well. Assuming you'd use this header map to put the `reactor_rabbitmq_delivery_tag` into that is a bit of a problem. 

Wouldn't it be better to have an BasicProperties (including the headers map) instantiated by default?

I also noticed this TODO in OutboundMessage `
`    // TODO add a correlation property to use for OutboundMessageResult
    // (instead of using the whole message)`

Maybe his correlation property could be used instead somehow? Although not entirely sure how we could then retrieve it in the returnListener's `handleReturn(int replyCode, String replyText, String exchange, String routingKey, AMQP.BasicProperties properties, byte[] body)`

**Comment by acogoluegnes** [comment]
@vanseverk `OutboundMessage#properties` is null by default because `Channel#basicPublish` accepts null properties.

I think we can leave the `OutboundMessage` instance the same, just add the header when publishing: https://github.com/reactor/reactor-rabbitmq/blob/edc3937a036d88d10871475329e3bd13c4aae0e0/src/main/java/reactor/rabbitmq/Sender.java#L635-L640 

This would look like the following in pseudo-code:

```
if (message.getProperties() == null) {
  ... create properties with delivery tag in `reactor_rabbitmq_delivery_tag` header
  channel.basicPublish(
    message.getExchange(),
    message.getRoutingKey(),
    newProps,
    message.getBody()
  );
} else {
  ... create new properties from messages properties, add delivery tag in `reactor_rabbitmq_delivery_tag` header
  channel.basicPublish(
    message.getExchange(),
    message.getRoutingKey(),
    newProps,
    message.getBody()
  );
}
```

This way we leave the `OutboundMessage` intact, the delivery tag trick is just an implementation detail. Note creating a copy of existing properties will be a bit tedious, because the properties will have to be copied one by one.

Note also this should be done in `ConfirmSendContext` as well (used to retry in case of error). This may be a bit trickier and is just polishing, I can do it once your PR is merged.

Forget about the TODO, it must be an old leftover.

The `reactor_rabbitmq_delivery_tag` header is a correlation ID as well, it should show up in the properties argument of `ReturnListener#handleReturn` and be used to clean up the unconfirmed map.

**Comment by a701440** [comment]
BTW does this mean that undeliverable messages would be causing a memory leak now? 
Given that they are not removed from that map? Are they removed from the map at some point?

**Comment by acogoluegnes** [comment]
@a701440 This could be a problem currently indeed.

**Comment by acogoluegnes** [comment]
@a701440 My bad, unroutable messages are confirmed as well, so there should be no memory leak. When publishing with mandatory, the [return listener is notified before the confirm listener](https://www.rabbitmq.com/confirms.html#when-publishes-are-confirmed), so this does not call this feature into question. We just have to be careful that the unconfirmed data structure may not contain a confirmed message because it has been returned already.

/cc @vanseverk

**Comment by a701440** [comment]
In this case why even bother about removing it from the map at the "return" stage? Let the "confirm" stage do the removal as it does now.

**Comment by acogoluegnes** [comment]
Good point :) But the return listener can emit downstream that the message is not routable, which is a useful information. Then when the message is confirmed, there's no way to know that something has already been emitted downstream for it, so another "confirmation" would be emitted.

I don't see any other way than emitting and cleaning the map in the return listener and let the confirm listener do nothing when it cannot find a confirmed message in the map.

**Comment by vanseverk** [comment]
@acogoluegnes in that case I'll have it deleted (and returned) through the returnlistener, to prevent the duplicate confirmation. Code for that should be okay now - still have to write a test for it though.

**Comment by acogoluegnes** [comment]
@vanseverk Sounds good. The current code in the confirm listener is not super defensive:

https://github.com/reactor/reactor-rabbitmq/blob/edc3937a036d88d10871475329e3bd13c4aae0e0/src/main/java/reactor/rabbitmq/Sender.java#L599-L605

`outboundMessage` should be checked for null to avoid an error in case of an already-returned message.

**Comment by vanseverk** [comment]
@acogoluegnes to continue our ping-pong some additional questions :-)

When the ReturnListener gets triggered, the OutboundMessageResult expects the "acknowledged" flag. I'm currently assuming it's not always true or false, however, so I guess we can't determine it in the returnListener? Because in that case, things definitely get more complicated for our scenario.

Is it currently actually possible to add the mandatory or immediate fields when sending a message? I assume these flags need to be added as well to support this further.

**Comment by acogoluegnes** [comment]
@vanseverk I think it's good enough to create an `OutboundMessageResult` instance with `acknowledged` set to `true` (from the `ReturnListener`). Returned messages are supposed to be acknowledged just after, so this is what makes sense I'd say. `OutboundMessageResult` should have a new `returned` flag, Javadoc should specify that the `acknowledged` flag makes less sense as soon as `returned` is set to `true`.

`SendOptions` should have a new `trackReturtned` flag and publishing should use `mandatory = true` if this new flag is set to `true`. WDYT?

**Comment by vanseverk** [comment]
@acogoluegnes sure, I'll add that to the Javadoc.

The flag has been added (haven't commited any code yet though), but I'm not sure whether that should lead to a `mandatory = true`, couldn't it also imply an `immediate = true`? I'm not a RabbitMQ expert at all, so still a bit doubtful. If it makes most sense to you, I can just go for the `mandatory = true`.

**Comment by acogoluegnes** [comment]
Don't worry about the `immediate` flag, it's not longer used by RabbitMQ.

**Comment by acogoluegnes** [comment]
Fixed in #102.
